### ISSUE TITLE
Another Way of Ensembling with LSTM

### ISSUE THREAD

**Issue by akaniklaus** [titu1994/MLSTM-FCN#7]
@titu1994 What do you think of the following way of ensembling with LSTM? (Feeds CNN outputs in FCN to both GAP and LSTM, and concat their outputs before the final Linear or Softmax layer).

I have read some papers that has CNN followed by LSTM and came up with this version of FCN. I unfortunately doesn't have time and resources to experiment on it as detailed as you have done.

In addition to that, have you ever tried utilizing GRU instead of the LSTM module in your experiments?

Hence, I would love to hear your idea about this and maybe results if you would try this as well.

```
class out_layer(nn.Module):
    def __init__(self, in_channels, out_channels = 1, gap_size = 1200, cl = False, lstm_size = 0):
        super(out_layer, self).__init__()
        self.in_channels = in_channels
        self.lstm_size = lstm_size
        self.gap = nn.AvgPool1d(kernel_size = gap_size, padding = 1)
        if lstm_size > 0:
            self.lin = nn.Linear(in_channels * 2, out_channels)
            self.lstm = nn.LSTM(input_size = gap_size, hidden_size = lstm_size, batch_first = True)
        else:
            self.lin = nn.Linear(in_channels, out_channels)
        self.activation = nn.Softmax() if cl else lambda x: x
    def forward(self, x):
        gap = self.gap(x).view(-1, self.in_channels)
        if self.lstm_size > 0:
            out_lstm, _ = self.lstm(x)
            gap = torch.cat((gap, out_lstm[:,:,-1]), 1)
        return self.activation(self.lin(gap))
```

**Comment by titu1994** [comment]
As to utilizing GRU instead of LSTM, early results showed that the GRU obtains worse performance on the 10-12 datasets we randomly tried it on. Perhaps it needs better tuning of hyper parameters. I believe it is simply under capacity. 

Thanks for the implementation of your code. The idea is interesting, but right now we do not have the resources to try it out on all 85 datasets. 

What I can guess is, (and this may be proven wrong), is that using CNN features as input to the lstm will probably cause overfitting in the UCR datasets. I have seen them work in image domains with tons of data, but for UCR, which has a minimum of 20 samples for the train set, it's too small to train such a model. 

However, we won't know until we try. I suggest experimenting a bit with Google Colab. You can use git to pull this entire repo, and link the data from your drive to the data directory, after which our scripts should work as normal python script calls.

**Comment by akaniklaus** [comment]
Here, some papers about CNN+LSTM usage (in other domains, not in time series classification):
https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43455.pdf
https://arxiv.org/pdf/1411.4389v3.pdf

These are some applications to sensor data I guess, for the purpose of action recognition:
http://www.mdpi.com/1424-8220/18/2/679/pdf
https://umu.diva-portal.org/smash/get/diva2:1161826/FULLTEXT01.pdf
https://scholarcommons.usf.edu/cgi/viewcontent.cgi?article=8032&context=etd

**Comment by titu1994** [comment]
The CNN does not perform dim shuffling, so the output of the CNN filters would be (N, T, 128), which is far larger than the current input of (N, 1, T) to the lstm. 

Yes I removed bias to reduce the number of parameters in Squeeze and Excitation blocks. 

You will find that when you run on all the experiments, overfitting is the cause for severe performance drops. Therefore every thing you can do to avoid that helps. 

You can try GRU on many datasets And see the overall results. I've seen ConvLSTMs being used before for various things, but I don't believe it is that useful here.

**Comment by akaniklaus** [comment]
@titu1994 Hello again, would you be able to test it only maybe in few (1-2) datasets where it is less likely to have an overfitting problem?
